### PR TITLE
 nixos/logrotate: switch `paths` option type from listOf to attrsOf

### PR DESCRIPTION
**Pull request by aanderse**
###### Motivation for this change
https://github.com/NixOS/nixpkgs/pull/87702

It was a really bad idea for me to suggest @jslight90 to use a `listOf` instead of `attrsOf` for the `paths` option. I upgraded a server to `unstable` and tried this out... `attrsOf` is a **much** better fit when given a `priority` option to control sort order.

With this PR NixOS modules can actually start providing log rotation for (the very few) services that actually need it using `logrotate`. The user remains in full control with the ability to even disable log rotation via the `enable` option, like so:
```
services.logrotate.paths.httpd.enable = false;
```

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by aanderse** [comment]
@Infinisil we had a small conversation about this module and I always appreciate/enjoy your reviews... would you mind taking a look at this?

**Review by infinisil — CHANGES_REQUESTED** [review]

File: nixos/modules/services/logging/logrotate.nix [hunk 1]
```diff
@@ -65,41 +106,63 @@ in
...
-        description = "List of attribute sets with paths to rotate";
-        example = {
-          "/var/log/myapp/*.log" = {
-            user = "myuser";
-            group = "mygroup";
-            rotate = "weekly";
-            keep = 5;
-          };
-        };
+        type = with types; attrsOf (submodule pathOpts);
```
[the remark this answers is not in the document]
Comment (aanderse) [reply nixos/modules/services/logging/logrotate.nix:109]: This option hasn't been introduced in a stable release yet, so breaking change is fine.

Many options in `logrotate.conf` are single value without a key so I'm thinking using `settings` over `extraConfig` is more confusing than the value it brings. The `logrotate` format uses the last declared value in the case of duplicates so the compelling reasons to use an attribute set aren't there in my opinion.

File: nixos/modules/services/logging/logrotate.nix [hunk 2]
```diff
@@ -65,41 +106,63 @@ in
...
+        description = ''
+          Attribute set of paths to rotate.
+        '';
+        example = literalExample ''
+          {
+            httpd = {
+              path = "/var/log/httpd/*.log";
+              user = config.services.httpd.user;
+              group = config.services.httpd.group;
+              keep = 7;
```
[the remark this answers is not in the document]
Comment (aanderse) [reply nixos/modules/services/logging/logrotate.nix:120]: Good point. Will do.

File: nixos/modules/services/logging/logrotate.nix [hunk 3]
```diff
@@ -65,41 +106,63 @@ in
...
     };
   };
 
   config = mkIf cfg.enable {
-    systemd.services.logrotate = {
-      description   = "Logrotate Service";
-      wantedBy      = [ "multi-user.target" ];
-      startAt       = "*-*-* *:05:00";
+    assertions = mapAttrsToList (name: pathOpts:
+      { assertion = pathOpts.user != null || pathOpts.group != null -> pathOpts.user != null && pathOpts.group != null;
```
[the remark this answers is not in the document]
Comment (aanderse) [reply nixos/modules/services/logging/logrotate.nix:147]: Thanks.

**Review by infinisil — APPROVED** [review]
The first commit looks good now, but I can't really comment much on the second one as I don't know a lot about httpd. If you think the second one won't cause any problems then feel free to merge it :)

**Review by ryantm — APPROVED** [review]
Looks good to me. Thanks!

**Pull request merged**